### DISCUSSION TITLE
Add or use the default action URL for Gitea / Forgejo Actions

### DISCUSSION THREAD

**Discussion by MichaelSasser** [renovatebot/renovate#31666] [Suggest an Idea]
### Tell us more.

By default, `uses: actions/checkout@v4` in Gitea/Forgejo Actions would tell the actions runner to download the action from github.com/actions/checkout. Administrators can change this behavior in the Gitea/Forgejo config file. 

For example, the [Configuration Cheat Sheet - actions (Actions) - DEFAULT_ACTIONS_URL](https://forgejo.org/docs/latest/admin/config-cheat-sheet/#actions-actions) contains the configuration:
> DEFAULT_ACTIONS_URL: https://code.forgejo.org/: Default address to get action plugins, e.g. the default value means downloading from “https://code.forgejo.org/actions/checkout” for “uses: actions/checkout@v3”

(Also see [Gitlab: Compared to GitHub Actions - Downloading actions](https://docs.gitea.com/usage/actions/comparison#downloading-actions))

By altering the configuration like `FORGEJO__actions__DEFAULT_ACTIONS_URL=self` the actions runner would download `actions/checkout` from the previous example from my own Gitea/Forgejo instance instead.

Currently, renovate doesn't seem to support an option where I can mirror this behavior, leading to to errors, when the action has a different than the one hosted on GitHub or when dealing with custom actions.

<details><summary>Example Dashboard with errors</summary>
This issue lists Renovate updates and detected dependencies. Read the [Dependency Dashboard](https://docs.renovatebot.com/key-concepts/dashboard/) docs to learn more.

## Repository problems

Renovate tried to run on this repository, but found these problems.

 - WARN: Package lookup failures
 - WARN: Attempting to gracefully recover from 409 Conflict response in createPr(chore(deps): update renovate/renovate docker tag to v38.100.0, renovate/renovate-renovate-38.x)

## Errored

These updates encountered an error and will be retried. Click on a checkbox below to force a retry now.

 - [ ] chore(deps): update renovate/renovate docker tag to v38.100.0

---

> ⚠️ **Warning**
> 
> Renovate failed to look up the following dependencies: `Failed to look up github-tags package actions/yamllint`.
> 
> Files affected: `.forgejo/workflows/renovate.yml`

---

## Detected dependencies

<details><summary>github-actions</summary>
<blockquote>

<details><summary>.forgejo/workflows/cleanup.yml</summary>

</details>

<details><summary>.forgejo/workflows/renovate.yml</summary>

 - `actions/checkout v4`
 - `actions/yamllint v3`
 - `actions/checkout v4`
 - `actions/checkout v4`
 - `ghcr.io/renovatebot/renovate 38.100.0`
 - `ghcr.io/renovatebot/renovate 38.100.0`

</details>

</blockquote>
</details>

<details><summary>woodpecker</summary>
<blockquote>

<details><summary>.woodpecker/cleanup.yml</summary>

 - `docker 27.3.1-cli`

</details>

<details><summary>.woodpecker/lint.yml</summary>

 - `pipelinecomponents/yamllint 0.32.1`

</details>

<details><summary>.woodpecker/renovate.yml</summary>

 - `renovate/renovate 38.98.1`

</details>

<details><summary>.woodpecker/validate.yml</summary>

 - `renovate/renovate 38.98.1`

</details>

</blockquote>
</details>
</details>

As a workaround, I think absolute URLs currently work `uses: https://my-gitea.com/actions/checkout@v4` but I haven't tested it yet

**Comment by viceice** [comment]
override registry url via package rules

**Reply by MichaelSasser** [reply]
Thanks for your swift response.

So far, I've tried it with `gihub-tags` and `gitea-tags` independently of each other and both at once. From the logs, it seems to always use the `github-tags` datasouce. I haven't seen anything about `gitea-tags`. 
This means, either way, I still get warnings like: `Failed to look up github-tags package actions/yamllint` in the logs.

The action `actions/yamllint` is a mirror of [ibiqlik/action-yamllint](https://github.com/ibiqlik/action-yamllint.git), which exists on my forgejo instance and is public (just like the `actions` organization). I can clone it without any issues.

```shell
# Inside some fresh OCI container with git installed
$ git clone 'https://my-forgejo-instance.foo.bar/actions/yamllint.git' /tmp/yamllint
Cloning into '/tmp/yamllint'...
remote: Enumerating objects: 188, done.
remote: Counting objects: 100% (188/188), done.
remote: Compressing objects: 100% (81/81), done.
remote: Total 188 (delta 88), reused 187 (delta 88), pack-reused 0 (from 0)
Receiving objects: 100% (188/188), 33.64 KiB | 6.73 MiB/s, done.
Resolving deltas: 100% (88/88), done.
```

Here is how I tried to override the registry URL(s):

```js
// config.js
module.exports = {
  // ...
  packageRules: [
    {
      matchDatasources: ["gitea-tags"],
      registryUrls: ["https://my-forgejo-instance.foo.bar"],
    },
    {
      matchDatasources: ["gihub-tags"],
      registryUrls: ["https://my-forgejo-instance.foo.bar"],
    },
  ],
  // ...
}
```

Those are the logs, when I only override the `gitea-tags` datasource or none at all. If I overrode the `gihub-tags` datasource, it would show the same warning for all actions (some of them, like `actions/checkout` have the same organization and name on GH and my instance).

```log
 # Logs
 INFO: Repository started (repository=renovate/renovate)
       "renovateVersion": "38.101.1"
 INFO: Dependency extraction complete (repository=renovate/renovate, baseBranch=main)
       "stats": {
         "managers": {
           "github-actions": {"fileCount": 1, "depCount": 8},
           "woodpecker": {"fileCount": 4, "depCount": 5}
         },
         "total": {"fileCount": 5, "depCount": 13}
       }
 WARN: Package lookup failures (repository=renovate/renovate)
       "warnings": ["Failed to look up github-tags package actions/yamllint"],
       "files": [".forgejo/workflows/renovate.yml"]
 INFO: Repository finished (repository=renovate/renovate)
```

Is there maybe a way how I can force the `github-actions` manager ([docs](https://docs.renovatebot.com/modules/manager/github-actions/#supported-datasources); [github-actions sources](https://github.com/renovatebot/renovate/blob/8e9c8cb2db493996cd372d6afef035c46761cb03/lib/modules/manager/github-actions/index.ts)) to use `gitea-tags` as the datasouce instead of `github-tags`? Or just `gitea-tags` since I don't really need the others. 

I mean, something like:

```js
module.exports = {
  "github-actions": {
    supportedDatasources: ["gitea-tags"], // Not a valid config option
  },
}
```

Dunno if it would help, though.

**Reply by viceice** [reply]
you need to set datasource and registry url via package rule

**Reply by MichaelSasser** [reply]
I'm not sure, but I think that's what I did. Please correct me if I'm wrong.

While searching for a solution, I found a similar issue https://github.com/renovatebot/renovate/discussions/27734.
